app/controllers/application_controller.rb

```diff
@@ -323,4 +324,8 @@ def handle_unverified_request
     logger.info("unverified_request session:" + session.inspect)
     reset_session
   end
+
+  def store_request_in_thread
+    Thread.current[:request] = request
```

Comment:
One comment here.  Maybe instead of saving the whole request object, we could just save the data we need, under an well named key.  I feel like Thread.current[:request] could very well be something important we don't want to mess with. At least I would assume that when reading it out of context.  Like maybe  `Thread.current[:saved_server_url]="#{request.protocol}#{request.host_with_port}"` ?